doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,637 @@
+creating the datatype for the array. Here, I use ``np.dtype[mp.mpf]``
+to find the DType class (the notation is not part of this NEP),
+which is then instantiated with the desired parameter.
+This could also be written as ``MpfDType`` class::
+
+    >>> arr2 = np.array([1, 2, 3], dtype=np.dtype[mp.mpf](dps=100))
+    >>> print(arr1 + arr2)
+    array(['0.0', '2.0', '4.0'], dtype=mpf[dps=100])
+
+The ``mpf`` datatype finds that the result of the operation should be the higher
+precision one of the two, so uses a precision of 100.
+Furthermore, we should be able to define casting, for example as in::
+
+    >>> np.can_cast(arr1.dtype, arr2.dtype, casting="safe")
+    True
+    >>> np.can_cast(arr2.dtype, arr1.dtype, casting="safe")
+    False
+    >>> np.can_cast(arr2.dtype, arr2.dtype, casting="same_kind")
+    True
+
+Casting from float is a probably always at least a ``same_kind`` cast, but
+in general, it is not safe::
+
+    >>> np.can_cast(np.float64, np.dtype[mp.mpf](dps=4), casting="safe")
+    False
+
+Since the a float64 has a higer precision than the ``mpf`` datatype with
+``dps=4``.
+Alternatively, we can say that::
+
+    >>> np.common_type(np.dtype[mp.mpf](dps=5), np.dtype[mp.mpf](dps=10))
+    np.dtype[mp.mpf](dps=10)
+
+And possibly even::
+
+    >>> np.common_type(np.dtype[mp.mpf](dps=5), np.float64)
+    np.dtype[mp.mpf](dps=16)  # equivalent precision to float64 (I believe)
+
+since ``np.float64`` can be cast to a ``np.dtype[mp.mpf](dps=16)`` safely.
+
+
+Unit on the Datatype
+""""""""""""""""""""
+
+There are different ways to define Units, depending on how the internal
+machinery would be organized, one way is to have a single Unit datatype
+for every existing numerical type.
+This will be written as ``Unit[float64]``, the unit itself is part of the
+DType instance ``Unit[float64]("m")`` us a ``float64`` with meters attached::
+
+    >>> meters = np.array([1, 2, 3], dtype=np.float64) * unit.m  # meters
+    >>> print(meters)
+    array([1.0, 2.0, 3.0], dtype=Unit[float64]("m"))
+
+Note that units are a bit tricky, since it is debatable, whether::
+
+    >>> np.array([1.0, 2.0, 3.0], dtype=Unit[float64]("m"))
+
+should be valid syntax (coercing the float scalars without a unit to meters).
+Once the array is created, math will work without any issue::
+
+    >>> meters * 2 * unit.seconds
+    array([2.0, 4.0, 6.0], dtype=Unit[float64]("ms"))
+
+Casting is not valid from one unit to the other, but can be between different
+scales of the same dimensionality (although this may usually be "unsafe")::
+
+    >>> meters.astype(Unit[float64]("s"))
+    TypeError: Cannot cast meters to seconds.
+    >>> meters.astype(Unit[float64]("km"))
+    >>> meters.astype(meters.dtype.to_cgs())
+
+The above notation may be somewhat clumsy, in some cases, and functions
+could be used to convert things otherwise.
+There may be ways to make these more convenient, but those must be left
+for future discussions::
```

Comment:
I'm glad to see this come up, even if the eventual decision is not to support these method names.

One caution for deciding a set of method names is that it might make it difficult for existing unit libraries to migrate to using ufuncs, since they would either need to abandon their existing API, happen to have the exact same API you decide to use for NumPy, or decide to keep using an ndarray subclass or wrapper class.

One thing that at least at one point was in the realm of possibility was adding a way for ufuncs to define methods that then get attached to ndarray. I totally understand if the eventual decision is that it's too weird or surprising for a ufunc to be able to modify the ndarray API. However, that's just about the only way I can think of for it to be possible for all existing unit libraries to migrate to this functionality without introducing API breakages.

Reply:
Yes, it may be possible to make the quantity subclass nothing (or little) more than a MixIn.

For a dtype to modify the ndarray seems weird, but I have not figured out how weird I think it is (I understand pandas to uses the dtype to indicate the correct subclass). While I like the dtype to be super central, the tuple `(ndarray, DType)` defines the actual theoretical type (space of instances/values) that we are dealing with. My vision is that we want to approach a world where `(Sparse, DType)` is plausible. That implies `DType` must know nothing about `ndarray` explicitly and prefers unit to be a `DType`.
But: how `(ndarray, DType)` *looks* from python is a slightly different discussion, maybe!

From that standpoint, you could think that a `DType` provides a MixIn which is always used and induces(?)  a subclass of `ndarray`. (Theoretically `(ndarray, np.dtype[float64])` *is* a subtype, so why not a subclass.) Or, you could also have a `ndarray.elementwise.method` namespace that is dtype dependent.

I love disucussing it, but for now I think it should not be in here since it is a "next step", fully independent of the other thoughts.